### PR TITLE
Per tfm PackageSpecOperations AddOrUpdate actually updates if the references is already present for that tfm. 

### LINKED ISSUE #6189: dotnet add no longer updates the package
## Details about Problem

NuGet product used (NuGet.exe | VS UI | Package Manager Console | dotnet.exe):
  - dotnet.exe 

dotnet.exe --version (if appropriate):
  - 2.0.2

OS version (i.e. win10 v1607 (14393.321)):
  - win10 v1607 (14393.1770)

Worked before? If so, with which NuGet version:
  - worked with dotnet.exe version 2.0.2 
## Detailed repro steps so we can see the same problem

1.  In a folder with a .csproj file with an out of date nuget package (ex: Newtonsoft.Json version 10.0.2) run command dotnet add package Netwonsoft.Json 

2.  Should automatically try to update package and restore nuget

...

#### Logs from dotnet.exe version 2.0.0 (EXPECTED)
PS C:\sampleProject\sampleProject> dotnet add package newtonsoft.json
  Writing C:\Users\millrya\AppData\Local\Temp\tmp3753.tmp
info : Adding PackageReference for package 'newtonsoft.json' into project 'C:\sampleProject\sampleProject\sampleProject.csproj'.
log  : Restoring packages for C:\sampleProject\sampleProject\sampleProject.csproj...
info :   CACHE http://proget.chrobinson.com/nuget/Pre-release/FindPackagesById()?id='Newtonsoft.Json'
info :   CACHE http://proget.chrobinson.com/nuget/Release/FindPackagesById()?id='Newtonsoft.Json'
info :   CACHE https://api.nuget.org/v3-flatcontainer/newtonsoft.json/index.json
info : Package 'newtonsoft.json' is compatible with all the specified frameworks in project 'C:\sampleProject\sampleProject\sampleProject.csproj'.
info : PackageReference for package 'newtonsoft.json' version '10.0.3' updated in file 'C:\sampleProject\sampleProject\sampleProject.csproj'.
PS C:\sampleProject\sampleProject>

#### Logs from dotnet.exe version 2.0.2 (ACTUAL)
PS C:\sampleProject\sampleProject> dotnet add package newtonsoft.json
  Writing C:\Users\millrya\AppData\Local\Temp\tmp3A42.tmp
info : Adding PackageReference for package 'newtonsoft.json' into project 'C:\sampleProject\sampleProject\sampleProject.csproj'.
log  : Restoring packages for C:\sampleProject\sampleProject\sampleProject.csproj...
info : Package 'newtonsoft.json' is compatible with all the specified frameworks in project 'C:\sampleProject\sampleProject\sampleProject.csproj'.
info : PackageReference for package 'newtonsoft.json' version '10.0.2' updated in file 'C:\sampleProject\sampleProject\sampleProject.csproj'.
PS C:\sampleProject\sampleProject>

### Sample Project

[sampleProject.zip](https://github.com/NuGet/Home/files/1475935/sampleProject.zip)

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug
Fixes: https://github.com/NuGet/Home/issues/6189 and   https://devdiv.visualstudio.com/DevDiv/_workitems?id=521955
Regression: No  
If Regression then when did it last work:  15.3
If Regression then how are we preventing it in future:   Adding more tests

## Fix
Details: The add or update method call in the package spec operations, was not checking whether the reference already existed before adding it to the package spec. Later the resolver would just get the first version of the package references in the package spec. 

## Testing/Validation
Tests Added: Yes  
Reason for not adding tests:  
Validation done:  Manual, Automated tests|

I will add more tests, but I'd like to get the core logic reviewed ASAP.

### CODE REVIEW

File: test/NuGet.Core.Tests/NuGet.ProjectModel.Test/PackageSpecOperationsTests.cs [hunk 1]
```diff
@@ -11,6 +12,82 @@ namespace NuGet.ProjectModel.Test
 {
     public class PackageSpecOperationsTests
     {
+
+        [Fact]
+        public void AddOrUpdateDependency_AddsNewPackageDependencyToAllFrameworks()
```
Comment (emgarten) [inline test/NuGet.Core.Tests/NuGet.ProjectModel.Test/PackageSpecOperationsTests.cs:17]: AllFrameworks isn't very fitting for a test with one framework :smile:

**Review by emgarten — APPROVED** [review]

**Review by jainaashish — APPROVED** [review]
Good unit tests, but I would love to see some E2E tests as well for this scenario.

File: test/NuGet.Core.Tests/NuGet.ProjectModel.Test/PackageSpecOperationsTests.cs (hunk 1 above)
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.ProjectModel.Test/PackageSpecOperationsTests.cs:17]: This is where an upside down shuffle parrot is appropriate :D

**Comment by nkolev92** [comment]
@jainaashish Yeah, looking into it, just having lots of trouble getting apex to cooperate. 

I have added more scenario tests for NuGet/Home#6189 though. 
The current tests are a bit messy, so I "patched" them up a bit.

**Review by mishra14 — APPROVED** [review]
🏅

**Pull request merged**